### PR TITLE
Update build environment and dependencies

### PR DESCRIPTION
**Pull request by NickCao**
Only tested power control and storage switching.

### CODE REVIEW

**Comment by mangelajo** [comment]
Nice work 🚀

File: application/src/control.rs [hunk 1]
```diff
@@ -0,0 +1,387 @@
...
+                    match key {
+                        ConfigKey::Name => {
+                            xfer.accept_with(&cfg.name).ok();
+                        }
+                        ConfigKey::Tags => {
+                            xfer.accept_with(&cfg.tags).ok();
+                        }
+                        ConfigKey::Json => {
+                            xfer.accept_with(&cfg.json).ok();
+                        }
```
Comment (mangelajo) [inline application/src/control.rs]: Just a note: no action is required.

We will have an issue with .json once it gets bigger than the transfer size. I don't know what the default is or what we configured. 

On the software I was starting to move name/tags/usbconsole under json , as those are details that the microcontroller itself does not care about, or use for anything,.... but the transfer size could be a good reason to stop doing such thing.

File: application/src/control.rs [hunk 2]
```diff
@@ -0,0 +1,387 @@
...
+                        ReadKey::Voltage => {
+                            let mut buf = heapless::Vec::<u8, 128>::new();
+                            write!(buf, "{:.2}V", self.data.voltage).ok();
+                            xfer.accept_with(&buf).ok();
+                        }
+                        ReadKey::Current => {
+                            let mut buf = heapless::Vec::<u8, 128>::new();
+                            write!(buf, "{:.2}A", self.data.current).ok();
+                            xfer.accept_with(&buf).ok();
+                        }
```
Comment (mangelajo) [inline application/src/control.rs]: Can we add another command to read them all at once?
I would expect that we periodically poll during the execution of a test, and ideally we would want the three values (actually just current and voltage would give us all we need), May be ReadKey:CurrentVoltage?

File: application/src/control.rs [hunk 3]
```diff
@@ -0,0 +1,387 @@
...
+            _ => return,
+        }
+
+        match req.request.try_into() {
+            Ok(ControlRequest::Nop) => {
+                xfer.accept().unwrap();
+            }
+            Ok(ControlRequest::Power) => {
+                if let Ok(action) = req.value.try_into() {
+                    self.power = Some(action);
```
Comment (mangelajo) [inline application/src/control.rs]: ah nice, so you decouple the reception of the control commands/data, with the actuator. I guess it isolates the objects each task has access to.

File: application/src/main.rs [hunk 4]
```diff
@@ -262,105 +252,49 @@ mod app {
...
 
-    #[task(binds = USART1, priority=1, local = [usart_rx, to_host_serial], shared = [shell_status, led_rx])]
+    #[task(binds = USART1, priority=1, local = [usart_rx], shared = [led_rx, serial2])]
     fn usart_task(cx: usart_task::Context){
         let usart_rx = cx.local.usart_rx;
-        let shell_status = cx.shared.shell_status;
-        let led_rx = cx.shared.led_rx;
-        let to_host_serial = cx.local.to_host_serial;
+        let led_rx   = cx.shared.led_rx;
+        let serial2  = cx.shared.serial2;
```
Comment (mangelajo) [inline application/src/main.rs]: may be now we should just call it dut_usb_serial or dut_vcp? there is no serial1

File: application/src/main.rs [hunk 5]
```diff
@@ -514,33 +404,12 @@ mod app {
...
-                    }
+                        usart_tx.write(c).ok();
+                } else {
+                    break
                 }
             }
         }
     }
 
-    fn escaped_char(c:u8) -> Option<u8> {
```
Comment (mangelajo) [inline application/src/main.rs]: so much complexity went away :D ... ❤️

**Comment by NickCao** [comment]
Extracted the control interface to its own PR: https://github.com/jumpstarter-dev/dutlink-firmware/pull/3

**Comment by mangelajo** [comment]
Ooppss sorry, now this one is in merge conflict, I must have merged this long ago

**Comment by NickCao** [comment]
Rebased, lemme test.

File: bootloader/src/main.rs [hunk 6]
```diff
@@ -288,14 +288,17 @@ fn dfu_init() -> LedType {
     /* USB device */
 
     let usb_dev = UsbDeviceBuilder::new(&bus, UsbVidPid(0x2b23, 0x1012))
-        .manufacturer("Red Hat Inc.")
-        .product("Jumpstarter DFU Mode")
-        .serial_number(get_serial_str())
+        .strings(&[
+            StringDescriptors::new(LangID::EN)
+            .manufacturer("Red Hat Inc.")
+            .product("Jumpstarter DFU Mode")
```
Comment (mangelajo) [inline bootloader/src/main.rs:294]:
```suggestion
            .product("Dutlink DFU Mode")
```

not related to this change, but since we are at it...

**Review by mangelajo — APPROVED** [review]
I have suggestion, but let's do it later :)

**Pull request merged**